### ISSUE TITLE
Custom Acknowledgement

### ISSUE THREAD

**Issue by keletsomolefe** [shamblett/mqtt_client#154]
How can I manually acknowledge QOS 1 messages? The current setup uses client.updates.listen(), what happens if the app crushes, do I lose all messages inside the listen() callback if the business logic is not applied on them?

**Comment by shamblett** [comment]
Currently this is the case, the client deposits the publish message on the update stream then sends an ack back to the broker, if the ack gets to the broker then the app crashes before it has been taken off the updates stream the publish message will not be sent again by the broker as it has been acked.

I can see why you may only want the ack to occur when you are sure you have processed the publish message. I'll have a look at putting a manual ack mechanism in the client for those who need this post release 6.0.0.

**Comment by keletsomolefe** [comment]
Java implementation https://www.eclipse.org/paho/files/javadoc/org/eclipse/paho/client/mqttv3/MqttCallback.html uses a method called messageReceived

Javascript implementation https://github.com/mqttjs/MQTT.js/blob/master/README.md#handleMessage uses handleMessage callback

**Comment by Husnain-Asharf** [comment]
hi any progress on this?

**Comment by shamblett** [comment]
Not as such, however I have some time to look at this now.

**Comment by shamblett** [comment]
You don't just have client.updates.listen, you also have this -
```
 /// If needed you can listen for published messages that have completed the publishing
  /// handshake which is Qos dependant. Any message received on this stream has completed its
  /// publishing handshake with the broker.
  client.published!.listen((MqttPublishMessage message) {
    print(
        'EXAMPLE::Published notification:: topic is ${message.variableHeader!.topicName}, with Qos ${message.header!.qos}');
    if (message.variableHeader!.topicName == topic3) {
      print('EXAMPLE:: Non subscribed topic received.');
    }
  });
```
from the mqtt_client_publish_qos1 .dart example.

Could you have a look at this example and see if this fulfils your needs?

**Comment by shamblett** [comment]
Forget the comment above, getting confused with the process for checking if a sent message is complete, not received ones like you need.

I've added a flag named manuallyAcknowledgeQos1 and a method named acknowledgeQos1Message, please have a look at the API docs for these, the client should now allow you to listen for a message, do any business logic processing on it, then manually ack it so if you crash before the ack is sent the broker will send it again. There is an example of usage in the examples directory namedmqtt_client_publish_qos1_manually_acknowledge.dart.

The package has been re published at version 9.2.0, please see if this is OK for you.
